bundles/org.openhab.ui.basic/src/main/java/org/openhab/ui/basic/internal/render/InputRenderer.java

```diff
@@ -58,38 +69,127 @@ public boolean canRender(Widget w) {
+
+        String inputType = "text";
+        String inputPattern = "";
+        if ("number".equals(inputHint)) {
+            inputType = "number";
+            inputPattern = "pattern=\"[0-9]*([\\.|,][0-9]*)?\"";
+        } else if ("date".equals(inputHint)) {
+            inputType = "date";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2}\"";
+        } else if ("time".equals(inputHint)) {
+            inputType = "time";
+            inputPattern = "pattern=\"[0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        } else if ("datetime".equals(inputHint)) {
+            inputType = "datetime-local";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2} [0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        }
+        snippet = snippet.replace("%input_type%", inputType);
+        snippet = snippet.replace("%input_pattern%", inputPattern);
+
+        String displayState = getValue(w);
+        State state = itemUIRegistry.getState(w);
+
+        String undefState = "-";
+        if ("datetime".equals(dataType)) {
+            undefState = "";
+            if ((inputHint == null || inputHint.isEmpty())) {
+                undefState = "YYYY-MM-DD hh:mm:ss";
+            }
+        } else if (state == null || state instanceof UnDefType) {
+            undefState = displayState;
+            if ("number".equals(inputHint)) {
+                String[] stateArray = displayState.trim().split(" ");
+                if (stateArray.length > 0) {
+                    undefState = stateArray[0];
+                }
+            }
+        }
+        snippet = snippet.replace("%undef_state%", undefState);
+
+        String dataState = displayState;
+        if (state == null || state instanceof UnDefType) {
+            dataState = "";
+        } else {
+            dataState = displayState.isEmpty() ? state.toString() : displayState;
+            if ("number".equals(inputHint)) {
+                String[] stateArray = displayState.trim().split(" ");
+                if (stateArray.length > 0) {
+                    dataState = parseNumber(stateArray[0]);
+                }
+            } else if (state instanceof DateTimeType) {
+                dataState = displayState.isEmpty() ? ((DateTimeType) state).format("%1$tY-%1$tm-%1$td %1$tT")
+                        : displayState;
+                if ("date".equals(inputHint)) {
+                    dataState = ((DateTimeType) state).format("%1$tY-%1$tm-%1$td");
+                } else if ("time".equals(inputHint)) {
+                    dataState = ((DateTimeType) state).format("%1$tT");
+                } else if ("datetime".equals(inputHint)) {
+                    dataState = ((DateTimeType) state).format("%1$tY-%1$tm-%1$tdT%1$tT");
+                }
+            }
+        }
+        snippet = snippet.replace("%data_state%", dataState);
+
+        String unitSnippet = "";
+        if ("number".equals(inputHint)) {
+            String[] stateArray = displayState.trim().split(" ");
```

Comment:
You should consider item state when displaySratte is empty. 
It explains one issue I found

Reply:
Changed.